device-telemetry/app/com/microsoft/azure/iotsolutions/devicetelemetry/services/external/DiagnosticsClient.java

```diff
@@ -0,0 +1,110 @@
+        if(this.diagnosticsEndpointUrl == null || this.diagnosticsEndpointUrl.length() == 0) {
+            this.log.error("No diagnostics url given, cannot write to diagnostics");
+            this.canWriteToDiagnostics = false;
+        } else {
+            this.canWriteToDiagnostics = true;
+        }
+        this.diagnosticsMaxLogRetries = servicesConfig.getDiagnosticsConfig().getMaxLogRetries();
+    }
+
+    @Override
+    public CompletionStage<Void> logEventAsync(String eventName) {
+        Dictionary<String, Object> emptyTable = new Hashtable<>();
+        return CompletableFuture.runAsync(() -> this.logEvent(eventName, emptyTable));
+    }
+
+    @Override
+    public CompletionStage<Void> logEventAsync(String eventName, Dictionary<String, Object> eventProperties) {
+        return CompletableFuture.runAsync(() -> this.logEvent(eventName, eventProperties));
+    }
+
+    @Override
+    public boolean canWriteToDiagnostics() {
+        return this.canWriteToDiagnostics;
+    }
+
+    private void logEvent(String eventName, Dictionary<String, Object> eventProperties) {
+        DiagnosticsRequestModel requestModel = new DiagnosticsRequestModel(eventName, eventProperties);
+        int retryCount = 0;
+        boolean requestSucceeded = false;
+        while (!requestSucceeded && retryCount < this.diagnosticsMaxLogRetries) {
+            try {
+                this.sendPostRequest(toJson(requestModel)).toCompletableFuture().get();
+                requestSucceeded = true;
+            } catch (Exception e) {
+                retryCount++;
+                if (retryCount < this.diagnosticsMaxLogRetries) {
+                    log.warn("Failed to log to diagnostics service, " + (this.diagnosticsMaxLogRetries - retryCount) + " retries remaining");
+                } else {
+                    log.error("Failed to log to diagnostics service, reached max number of retries");
+                }
+            }
+        }
+    }
+
+    private CompletionStage<Object> sendPostRequest(JsonNode jsonData) {
+        return this.prepareRequest()
+                .post(jsonData.toString())
+                .handle((result, error) -> {
+                    if (error != null) {
+                        log.error("Error logging to diagnostics service {}",
+                            error.getMessage());
+                        throw new CompletionException(
+                            new ExternalDependencyException(error.getMessage()));
+                    }
+
+                    if (result.getStatus() != HttpStatus.SC_OK) {
+                        log.error("Error logging to diagnostics service {}",
+                            error.getMessage());
+                        throw new CompletionException(
+                            new ExternalDependencyException(result.getStatusText()));
+                    }
+                    return true;
+                });
+    }
+
+    private WSRequest prepareRequest() {
```

Comment:
missing 'id' parameter here?

Reply:
no we don't need an id for the request to diagnosticsevents